In [1]:
import sys
import os

sys.path
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)
sys.path


['',
 '/opt/anaconda3/lib/python36.zip',
 '/opt/anaconda3/lib/python3.6',
 '/opt/anaconda3/lib/python3.6/lib-dynload',
 '/opt/anaconda3/lib/python3.6/site-packages',
 '/opt/anaconda3/lib/python3.6/site-packages/IPython/extensions',
 '/home/desenfirman/.ipython',
 '/media/Data/desenfirman/development/som-clustering-knn-imputation']

In [3]:
from main_algorithm import self_organizing_maps as som
from main_algorithm import knn_imputation as knn
    

In [4]:
def silhouette_visualizer(trained_weight, dataset):
    silhouette_res = som.silhouette(trained_weight, dataset)
    import matplotlib.pyplot as plt
    import random as rd
    fig, ax = plt.subplots()
    
    x = list()
    y = list()
    
    cmap = list()
    
    count = 1
    for idj, cluster in silhouette_res.items():
        if idj != 'avg':
            random_color = (rd.random(), rd.random(), rd.random(), 0.25)
            for i in cluster:
                x.append(i[1])
                y.append(count)
                cmap.append(random_color)
                count += 1
                
    ax.barh(y, x)
    ax.barh(y, [1] * len(dataset), color=cmap)
    ax.get_yaxis().set_ticks([])
    ax.axvline(silhouette_res['avg'], ls='--', color='r')
    plt.text(silhouette_res['avg'], len(dataset) + 5, 'avg silhouette: ' + str(round(silhouette_res['avg'], 2)))
    title = 'Silhouette Result of ' + str(len(silhouette_res) - 1) + ' Cluster(s)\n\n'
    plt.title(title)
    plt.xlabel('Silhouette Score')
    plt.ylabel('Input dataset')
    plt.show()

In [5]:
# Load dataset to jupyter notebook
from pandas import read_csv
import random

def extract_dataset(file):
    dataset = read_csv(file)
    new_columns = dataset.columns.values
    new_columns[0] = 'First_Column'
    dataset.columns = new_columns
    dataset = dataset.set_index('First_Column')
    raw_dataset = dataset
    return raw_dataset

dataset = extract_dataset("../dataset_used/dataset_PMKS.csv")
dataset

# dataset_used = dataset.iloc[:, :].values
# dataset_knn_ed = knn.impute_dataset(dataset_used, 0)
# dataset_norm = som.normalize_data(dataset_knn_ed)

# attr_size = len(dataset_used[0])
# weight = som.init_som_net(3, 3, attr_size)
# trained_weight = som.training(dataset_norm, weight, 100, 0.8, 0.2)
# silhouette_visualizer(trained_weight, dataset_norm)


ANAK JALANAN  TUNA SUSILA  PENGEMIS  \
First_Column                                            
Pacitan                    NaN          NaN         6   
Ponorogo                  52.0          NaN       153   
Trenggalek                16.0          NaN        68   
Tulungagung                6.0        162.0        32   
Blitar                    46.0          NaN        88   
Kediri                   184.0        107.0       154   
Malang                   210.0          NaN       220   
Lumajang                  15.0        116.0        97   
Jember                   171.0        172.0       332   
Banyuwangi                 NaN          NaN       297   
Bondowoso                 84.0         50.0       100   
Situbondo                 10.0         20.0        20   
Probolinggo              431.0        136.0        89   
Pasuruan                 120.0         18.0        43   
Sidoarjo                 145.0         20.0       125   
Mojokerto                102.0         99.0       108   
Jombang                   16.0         71.0        71   
Nganjuk                   91.0          NaN        84   
Madiun                    93.0        122.0       152   
Magetan                   38.0         90.0         7   
Ngawi                     86.0         12.0       106   
Bojonegoro                36.0         61.0       153   
Tuban                     31.0         62.0       194   
Lamongan                  26.0         11.0        10   
Gresik                   112.0        122.0       232   
Bangkalan                 12.0          5.0        30   
Sampang                   36.0          1.0        14   
Pameksan                   NaN          NaN       174   
Sumenep                   19.0          NaN       114   
Kota Kediri               18.0          9.0         8   
Kota Blitar               15.0         71.0         4   
Kota Malang              104.0          4.0        71   
Kota Probolinggo           NaN          NaN        95   
Kota Pasuruan             18.0          NaN        74   
Kota Mojokerto             6.0          NaN        58   
Kota Madiun                3.0          4.0        33   
Kota Surabaya             50.0          5.0        62   
Kota Batu                  3.0          4.0         5   

                  GELANDANGAN DAN GELANDANGAN PSIKOTIK  ANAK BALITA TERLANTAR  \
First_Column                                                                    
Pacitan                                           32.0                    NaN   
Ponorogo                                         114.0                  562.0   
Trenggalek                                        42.0                 4471.0   
Tulungagung                                       13.0                   87.0   
Blitar                                            75.0                   94.0   
Kediri                                           132.0                    8.0   
Malang                                           139.0                   57.0   
Lumajang                                           6.0                   21.0   
Jember                                           265.0                  142.0   
Banyuwangi                                       267.0                  333.0   
Bondowoso                                        212.0                  100.0   
Situbondo                                          NaN                    0.0   
Probolinggo                                       76.0                  593.0   
Pasuruan                                          58.0                   58.0   
Sidoarjo                                         102.0                   81.0   
Mojokerto                                         62.0                 1031.0   
Jombang                                          179.0                   40.0   
Nganjuk                                           97.0                   48.0   
Madiun                                            14.0                  349.0   
Magetan                             

# Pengujian parameter nilai learning rate (alpha)

Nilai yang akan diujikan
- 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1

Variabel kontrol
- Nilai eta = 0.1
- Jumlah epoch = 200
- Jumlah neuron = 5x5
- Nilai K = 0

In [5]:
from IPython.display import clear_output
from time import sleep
from pandas import DataFrame as df

# value to be tested
alphas = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]

# parameter control
eta = 0.5
epoch = 200
jumlah_neuron = (5,5)
k = 0

n_test = 10

qes = dict()


dataset_used = dataset.iloc[:, :].values
dataset_knn_ed = knn.impute_dataset(dataset_used, k)
dataset_norm = som.normalize_data(dataset_knn_ed)

for alpha in alphas:
    sum_all = 0
    qe_n = dict()
    for n in range(n_test):
        attr_size = len(dataset_used[0])
        print("Alpha: {0}, Eta: {1}, Epoch: {2}, Jumlah Neuron:{3}, K: {4} Try: {5}".format(alpha, eta, epoch, jumlah_neuron, k, n+1), end='\n')
        weight = som.init_som_net(jumlah_neuron[0], jumlah_neuron[1], attr_size)
        trained_weight = som.training(dataset_norm, weight, epoch, alpha, eta)
        qe = som.average_silhouette(trained_weight, dataset_norm)
        print("\nAverage Silhouette: {0}".format(qe), end='\n', flush=True)
        qe_n[n + 1] = qe
        sum_all += qe
    qe_n['avg'] = sum_all / n_test
    qes[str(alpha)] = qe_n
    clear_output()

(df.from_dict(qes)).to_csv('silhouette_test/alpha_testing.csv')

Alpha: 0.1, Eta: 0.5, Epoch: 200, Jumlah Neuron:(5, 5), K: 0 Try: 1


KeyboardInterrupt: 

# Pengujian parameter nilai signifikasi update bobot (eta)

Nilai yang akan diujikan
- 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1

Variabel kontrol
- Nilai alpha = 1
- Jumlah epoch = 200
- Jumlah neuron = 5x5
- Nilai K = 0

In [ ]:
from IPython.display import clear_output
from time import sleep
from pandas import DataFrame as df

# value to be tested
etas = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]

# parameter control
alpha = 0.5
epoch = 200
jumlah_neuron = (5,5)
k = 0

n_test = 10

qes = dict()


dataset_used = dataset.iloc[:, :].values
dataset_knn_ed = knn.impute_dataset(dataset_used, k)
dataset_norm = som.normalize_data(dataset_knn_ed)

for eta in etas:
    sum_all = 0
    qe_n = dict()
    for n in range(n_test):
        attr_size = len(dataset_used[0])
        print("Alpha: {0}, Eta: {1}, Epoch: {2}, Jumlah Neuron:{3}, K: {4} Try: {5}".format(alpha, eta, epoch, jumlah_neuron, k, n+1), end='\n')
        weight = som.init_som_net(jumlah_neuron[0], jumlah_neuron[1], attr_size)
        trained_weight = som.training(dataset_norm, weight, epoch, alpha, eta)
        qe = som.average_silhouette(trained_weight, dataset_norm)
        print("\nAverage Silhouette: {0}".format(qe), end='\n', flush=True)
        qe_n[n + 1] = qe
        sum_all += qe
    qe_n['avg'] = sum_all / n_test
    qes[str(eta)] = qe_n

(df.from_dict(qes)).to_csv('silhouette_test/eta_testing.csv')

# Pengujian parameter jumlah epoch

Nilai yang akan diujikan
- 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200

Variabel kontrol
- Nilai alpha = 0.5
- Nilai eta = 0.5
- Jumlah neuron = 5x5
- Nilai K = 0

In [6]:
from IPython.display import clear_output
from time import sleep
from pandas import DataFrame as df

# value to be tested
epochs = [1, 3, 4, 7, 11, 18, 29, 47, 76, 123, 76, 47, 29, 18, 11, 7, 4, 3, 1]

# parameter control
alpha = 0.5
eta = 0.5
jumlah_neuron = (5,5)
k = 0

n_test = 10

qes = dict()


dataset_used = dataset.iloc[:, :].values
dataset_knn_ed = knn.impute_dataset(dataset_used, k)
dataset_norm = som.normalize_data(dataset_knn_ed)

for epoch in epochs:
    sum_all = 0
    qe_n = dict()
    for n in range(n_test):
        attr_size = len(dataset_used[0])
        print("Alpha: {0}, Eta: {1}, Epoch: {2}, Jumlah Neuron:{3}, K: {4} Try: {5}".format(alpha, eta, epoch, jumlah_neuron, k, n+1), end='\n')
        weight = som.init_som_net(jumlah_neuron[0], jumlah_neuron[1], attr_size)
        trained_weight = som.training(dataset_norm, weight, epoch, alpha, eta)
        qe = som.average_silhouette(trained_weight, dataset_norm)
        print("\nAverage Silhouette: {0}".format(qe), end='\n', flush=True)
        qe_n[n + 1] = qe
        sum_all += qe
    qe_n['avg'] = sum_all / n_test
    qes[str(epoch)] = qe_n

(df.from_dict(qes)).to_csv('silhouette_test/epoch_testing_rev.csv')

Alpha: 0.5, Eta: 0.5, Epoch: 1, Jumlah Neuron:(5, 5), K: 0 Try: 1
Progress: 100.0% AVG Silhouette Score:0.2560740712512858
Average Silhouette: 0.2560740712512858
Alpha: 0.5, Eta: 0.5, Epoch: 1, Jumlah Neuron:(5, 5), K: 0 Try: 2
Progress: 100.0% AVG Silhouette Score:0.2574029276499657
Average Silhouette: 0.2574029276499657
Alpha: 0.5, Eta: 0.5, Epoch: 1, Jumlah Neuron:(5, 5), K: 0 Try: 3
Progress: 100.0% AVG Silhouette Score:0.22271038566557821
Average Silhouette: 0.22271038566557821
Alpha: 0.5, Eta: 0.5, Epoch: 1, Jumlah Neuron:(5, 5), K: 0 Try: 4
Progress: 100.0% AVG Silhouette Score:0.13488378845505128
Average Silhouette: 0.13488378845505128
Alpha: 0.5, Eta: 0.5, Epoch: 1, Jumlah Neuron:(5, 5), K: 0 Try: 5
Progress: 100.0% AVG Silhouette Score:0.2602604304921621
Average Silhouette: 0.2602604304921621
Alpha: 0.5, Eta: 0.5, Epoch: 1, Jumlah Neuron:(5, 5), K: 0 Try: 6
Progress: 100.0% AVG Silhouette Score:0.2573288427289215
Average Silhouette: 0.2573288427289215
Alpha: 0.5, Eta: 0.5, Ep

Alpha: 0.5, Eta: 0.5, Epoch: 11, Jumlah Neuron:(5, 5), K: 0 Try: 9
Progress: 100.0% AVG Silhouette Score:0.13870631504066488315040664888
Average Silhouette: 0.13870631504066488
Alpha: 0.5, Eta: 0.5, Epoch: 11, Jumlah Neuron:(5, 5), K: 0 Try: 10
Progress: 100.0% AVG Silhouette Score:0.125982007285275450072852754554
Average Silhouette: 0.12598200728527545
Alpha: 0.5, Eta: 0.5, Epoch: 18, Jumlah Neuron:(5, 5), K: 0 Try: 1
Progress: 100.0% AVG Silhouette Score:0.100472781659178572781659178577
Average Silhouette: 0.10047278165917857
Alpha: 0.5, Eta: 0.5, Epoch: 18, Jumlah Neuron:(5, 5), K: 0 Try: 2
Progress: 100.0% AVG Silhouette Score:0.079981221249155721221249155722
Average Silhouette: 0.07998122124915572
Alpha: 0.5, Eta: 0.5, Epoch: 18, Jumlah Neuron:(5, 5), K: 0 Try: 3
Progress: 100.0% AVG Silhouette Score:0.086687647927981567647927981566
Average Silhouette: 0.08668764792798156
Alpha: 0.5, Eta: 0.5, Epoch: 18, Jumlah Neuron:(5, 5), K: 0 Try: 4
Progress: 100.0% AVG Silhouette Score:0.135

Progress: 100.0% AVG Silhouette Score:0.070861867882720888678827208888
Average Silhouette: 0.07086186788272088
Alpha: 0.5, Eta: 0.5, Epoch: 123, Jumlah Neuron:(5, 5), K: 0 Try: 6
Progress: 100.0% AVG Silhouette Score:0.127880918254267809182542678886
Average Silhouette: 0.1278809182542678
Alpha: 0.5, Eta: 0.5, Epoch: 123, Jumlah Neuron:(5, 5), K: 0 Try: 7
Progress: 100.0% AVG Silhouette Score:0.093949429958177014299581770111
Average Silhouette: 0.09394942995817701
Alpha: 0.5, Eta: 0.5, Epoch: 123, Jumlah Neuron:(5, 5), K: 0 Try: 8
Progress: 100.0% AVG Silhouette Score:0.061607467512015874675120158777
Average Silhouette: 0.06160746751201587
Alpha: 0.5, Eta: 0.5, Epoch: 123, Jumlah Neuron:(5, 5), K: 0 Try: 9
Progress: 100.0% AVG Silhouette Score:0.125041088643552230886435522333
Average Silhouette: 0.12504108864355223
Alpha: 0.5, Eta: 0.5, Epoch: 123, Jumlah Neuron:(5, 5), K: 0 Try: 10
Progress: 100.0% AVG Silhouette Score:0.0669552188603982421886039824444
Average Silhouette: 0.06695521886

Progress: 100.0% AVG Silhouette Score:0.12923343449249713434492497114
Average Silhouette: 0.1292334344924971
Alpha: 0.5, Eta: 0.5, Epoch: 11, Jumlah Neuron:(5, 5), K: 0 Try: 2
Progress: 100.0% AVG Silhouette Score:0.110830515186121965151861219664
Average Silhouette: 0.11083051518612196
Alpha: 0.5, Eta: 0.5, Epoch: 11, Jumlah Neuron:(5, 5), K: 0 Try: 3
Progress: 100.0% AVG Silhouette Score:0.144192891713342438917133424333
Average Silhouette: 0.14419289171334243
Alpha: 0.5, Eta: 0.5, Epoch: 11, Jumlah Neuron:(5, 5), K: 0 Try: 4
Progress: 100.0% AVG Silhouette Score:0.117729744848727137448487271333
Average Silhouette: 0.11772974484872713
Alpha: 0.5, Eta: 0.5, Epoch: 11, Jumlah Neuron:(5, 5), K: 0 Try: 5
Progress: 100.0% AVG Silhouette Score:0.118291652537831146525378311442
Average Silhouette: 0.11829165253783114
Alpha: 0.5, Eta: 0.5, Epoch: 11, Jumlah Neuron:(5, 5), K: 0 Try: 6
Progress: 100.0% AVG Silhouette Score:0.125115847835714558478357145505
Average Silhouette: 0.1251158478357145
Al

Progress: 100.0% AVG Silhouette Score:0.23266647289579476
Average Silhouette: 0.23266647289579476
Alpha: 0.5, Eta: 0.5, Epoch: 1, Jumlah Neuron:(5, 5), K: 0 Try: 10
Progress: 100.0% AVG Silhouette Score:0.23896946736496688
Average Silhouette: 0.23896946736496688


# Pengujian parameter nilai jumlah neuron

Nilai yang akan diujikan
- 2x2, 3x3, 4x4, 5x5

Variabel kontrol
- Nilai alpha = 1
- Nilai eta = 0.7
- Jumlah epoch = 600
- Nilai K = 0

In [ ]:
from IPython.display import clear_output
from time import sleep
from pandas import DataFrame as df
import json

# value to be tested
eta = 0.5

# parameter control
alpha = 0.5
epoch = 200
jumlah_neurons = [(2,2), (3,3), (4,4), (5,5), (6,6)]
k = 0

n_test = 10

qes = dict()


dataset_used = dataset.iloc[:, :].values
dataset_knn_ed = knn.impute_dataset(dataset_used, k)
dataset_norm = som.normalize_data(dataset_knn_ed)

def cluster_build(dataset_indexes, dataset_input, trained_weight):
    cluster_result = dict()
    cls_list = list()
    for idx, x_data in enumerate(dataset_input):
        indexes = dataset_indexes[idx]
        c = som.penentuan_cluster(trained_weight, x_data)
        cls_id = str(c[0]) + ';' + str(c[1]) + ';'
        if cls_id not in cls_list:
            cls_list.append(cls_id)
            cluster_result[cls_id] = list()
        cluster_result[cls_id].append(indexes)
    return cluster_result
    

for jumlah_neuron in jumlah_neurons:
    sum_all = 0
    qe_n = dict()
    max_qe = -999999
    for n in range(n_test):
        attr_size = len(dataset_used[0])
        print("Alpha: {0}, Eta: {1}, Epoch: {2}, Jumlah Neuron:{3}, K: {4} Try: {5}".format(alpha, eta, epoch, jumlah_neuron, k, n+1), end='\n')
        weight = som.init_som_net(jumlah_neuron[0], jumlah_neuron[1], attr_size)
        trained_weight = som.training(dataset_norm, weight, epoch, alpha, eta)
        qe = som.average_silhouette(trained_weight, dataset_norm)
        print("\nAverage Silhouette: {0}".format(qe), end='\n', flush=True)
        qe_n[n + 1] = qe
        sum_all += qe
        
#         if qe > max_qe:
#             max_qe = qe
#             cls_result = cluster_build(dataset.index, dataset_norm, trained_weight)
#             cls_result['score'] = max_qe
#             with open('neuron_test_result/' + str(jumlah_neuron) + 'sil.json', 'w') as file:
#                 file.write(json.dumps(cls_result))
    qe_n['avg'] = sum_all / n_test
    qes[str(jumlah_neuron)] = qe_n

(df.from_dict(qes)).to_csv('silhouette_test/jumlah_neuron_testing.csv')

# Pengujian parameter nilai K

Nilai yang akan diujikan
- 0, 1, 2, 3, 4, 5, 6

Variabel kontrol
- Nilai alpha = 0.5
- Nilai eta = 0.5
- Jumlah epoch = 200
- Jumlah neuron = 5x5
- Nilai K = 0

In [ ]:
from IPython.display import clear_output
from time import sleep
from pandas import DataFrame as df

# value to be tested
ks = [0, 1, 2, 3, 4, 5, 6]

# parameter control
alpha = 0.5
eta = 0.5
jumlah_neuron = (5,5)
epoch = 200

n_test = 10

qes = dict()


dataset_used = dataset.iloc[:, :].values

for k in ks:
    sum_all = 0
    qe_n = dict()
    dataset_knn_ed = knn.impute_dataset(dataset_used, k)
    dataset_norm = som.normalize_data(dataset_knn_ed)
    for n in range(n_test):
        attr_size = len(dataset_used[0])
        print("Alpha: {0}, Eta: {1}, Epoch: {2}, Jumlah Neuron:{3}, K: {4} Try: {5}".format(alpha, eta, epoch, jumlah_neuron, k, n+1), end='\n')
        weight = som.init_som_net(jumlah_neuron[0], jumlah_neuron[1], attr_size)
        trained_weight = som.training(dataset_norm, weight, epoch, alpha, eta)
        qe = som.average_silhouette(trained_weight, dataset_norm)
        print("\nAverage Silhouette: {0}".format(qe), end='\n', flush=True)
        qe_n[n + 1] = qe
        sum_all += qe
            
    qe_n['avg'] = sum_all / n_test
    qes[str(epoch)] = qe_n

(df.from_dict(qes)).to_csv('silhouette_test/k_testing.csv')

In [ ]:
a = dict()
len(a)